### PR TITLE
React Native Path from Manifest

### PR DESCRIPTION
**Pull request by prashant1207**
**Objective**

To support scenarios where there is custom react-native fork, we need to have configurable way to get react-native path.

**Changes**
We have added a new property in platform specific configuration in `manifest.json`, similar to `bundleIdentifier`.
```
{
  "name": "Example",
  "displayName": "Example",
  "components": [
    {
      "appKey": "Example",
      "displayName": "App"
    }
  ],
  "macos": {
    "bundleIdentifier": "..."
    "reactNativePath": "react-native-macos"
  }
}
```

If the **app manifest** has the property `reactNativePath`, we will use that as react-native path, else we fallback to default value.

### CODE REVIEW

File: ios/test_app.rb [hunk 1]
```diff
@@ -80,13 +92,18 @@ def project_path(file, target_platform)
...
-  react_native = case target_platform
-                 when :ios then 'react-native'
-                 when :macos then 'react-native-macos'
-                 else raise "Unsupported target platform: #{target_platform}"
-                 end
-  Pathname.new(resolve_module(react_native))
+def react_native_path(project_root, target_platform)
+  react_native_from_manifest = react_native_path_from_manifest(project_root, target_platform)
+  if !react_native_from_manifest.nil?
+    react_native_from_manifest
```
Comment (tido64) [inline ios/test_app.rb]: Prefer to return early:
```suggestion
  return react_native_from_manifest if react_native_from_manifest.is_a? String
```

File: ios/test_app.rb [hunk 2]
```diff
@@ -38,6 +38,18 @@ def bundle_identifier(project_root, target_platform)
...
 end
 
+def react_native_path_from_manifest(project_root, target_platform) 
+  manifest_path = find_file('app.json', project_root)
+  unless manifest_path.nil?
+      manifest = JSON.parse(File.read(manifest_path))
+      platform_config = manifest[target_platform]
+      if !platform_config.nil? && !platform_config.empty?
+        reactNativePath = platform_config['reactNativePath']
+        return reactNativePath if reactNativePath.is_a? String
```
Comment (tido64) [inline ios/test_app.rb:48]:
```suggestion
        return Pathname.new(reactNativePath) if reactNativePath.is_a? String
```

Comment (tido64) [inline ios/test_app.rb:48]: We need to resolve `reactNativePath` to a path relative to where `Podfile` lives.

File: ios/test_app.rb [hunk 3]
```diff
@@ -38,6 +38,18 @@ def bundle_identifier(project_root, target_platform)
   @test_app_bundle_identifier
 end
 
+def react_native_path_from_manifest(project_root, target_platform)
```
Comment (tido64) [inline ios/test_app.rb:41]: Can we add tests for this?

**Review by tido64 — CHANGES_REQUESTED** [review]

File: ios/test_app.rb [hunk 4]
```diff
@@ -38,6 +38,18 @@ def bundle_identifier(project_root, target_platform)
   @test_app_bundle_identifier
 end
 
+def react_native_path_from_manifest(project_root, target_platform) 
+  manifest_path = find_file('app.json', project_root)
+  unless manifest_path.nil?
+      manifest = JSON.parse(File.read(manifest_path))
+      platform_config = manifest[target_platform]
+      if !platform_config.nil? && !platform_config.empty?
```
[the remark this answers is not in the document]
Comment (prashant1207) [reply ios/test_app.rb:46]: ack.

File: ios/test_app.rb (hunk 2 above)
Comment (prashant1207) [reply ios/test_app.rb]: I added the path resolution using `Pathname.new(resolve_module(reactNativePath)) if reactNativePath.is_a? String`, would it not suffice?

File: ios/test_app.rb (hunk 3 above)
Comment (prashant1207) [reply ios/test_app.rb:46]: For adding positive cases for `react_native_path_from_manifest` will require changes in manifest, Do we use any mock manifest while running the test?

File: ios/test_app.rb (hunk 2 above)
Comment (tido64) [reply ios/test_app.rb]: Can you post what `Podfile.lock` looks like after running `pod install`? We need to make sure we don't get absolute paths in there.

File: ios/test_app.rb (hunk 3 above)
Comment (tido64) [reply ios/test_app.rb]: Yes, look under `/test/__fixtures__`

File: ios/test_app.rb (hunk 2 above)
Comment (prashant1207) [reply ios/test_app.rb]: Pushed the podfile.lock.

File: ios/test_app.rb (hunk 3 above)
Comment (prashant1207) [reply ios/test_app.rb]: Alright, got it. I will add the tests there.

Comment (prashant1207) [reply ios/test_app.rb]: Added test cases.

**Review by alloy — APPROVED** [review]
LGTM 👍

**Review by tido64 — APPROVED** [review]
Thanks, Prashant 👍

**Comment by tido64** [comment]
Wiki updated: https://github.com/microsoft/react-native-test-app/wiki/Manifest-(app.json)/_compare/c469860ab6beecc40e3f608dcd9fb13d5374483b...05b5c55227bd29aa41a44cfc078f38d091616fe2

**Pull request merged**